Add database configuration to configure `in_clause_length`

yahonda opened the pull request:
### Summary

MySQL 5.7 introduces `range_optimizer_max_mem_size` to control memory usage for range access method. It sometimes changes execution plans of SQL statements with a large number of in clause, like 10000 from index range scan to full table scan.

"Limiting Memory Use for Range Optimization"
https://dev.mysql.com/doc/refman/5.7/en/range-optimization.html#range-optimization-memory-use

> If the specified limit is about to be exceeded, the range access method is abandoned
> and other methods, including a full table scan, are considered instead.

Instead of changing `range_optimizer_max_mem_size` value at MySQL server,
setting `in_clause_length` in Rails splits large in clause into smaller in list.

As of right now, `in_clause_length` can be configured as follows.

* Quote from @kamipo blog article https://blog.kamipo.net/entry/2017/12/23/230514 in Japanese

```ruby
ActiveSupport.on_load(:active_record) do
  module ActiveRecord
    module ConnectionAdapters
      module DatabaseLimitsExt
        def in_clause_length
          1000
        end
      end

      class AbstractAdapter
        prepend DatabaseLimitsExt
      end
    end
  end
end
```

This commit enables to configure `in_clause_length` value in the database.yml file.

* database.yml

```
production:
  adapter: mysql2
  in_clause_length: 1000
```

`in_clause_length` has been useful for Oracle enhanced adapter since Oracle database has a limit to allow the number of in list to 1000.
This parameter is not limited to MySQL. I have not found use cases for PostgreSQL and SQLite databases, though.

kamipo commented:
As an author of the blog post, that is just a thought experiment.

Unfortunately, `in_clause_length` only works for preloading queries (and also only used for Oracle enhanced adapter), so it couldn't work/help for others (eager loading, `where(id: too_many_ids)`, etc) for now.

yahonda committed:
Add database configuration to configure `in_clause_length`

MySQL 5.7 introduces `range_optimizer_max_mem_size` to control memory usage for range access method.
It sometimes changes execution plans of SQL statements with a large number of in clause, like 10000 from index range scan to full table scan.

"Limiting Memory Use for Range Optimization"
https://dev.mysql.com/doc/refman/5.7/en/range-optimization.html#range-optimization-memory-use

> If the specified limit is about to be exceeded, the range access method is abandoned
> and other methods, including a full table scan, are considered instead.

Instead of changing `range_optimizer_max_mem_size` value at MySQL server,
setting `in_clause_length` in Rails splits large in clause into smaller in list.

As of right now, `in_clause_length` can be configured as follows.

* Quote from kamipo's blog article https://blog.kamipo.net/entry/2017/12/23/230514 in Japanese

```ruby
ActiveSupport.on_load(:active_record) do
  module ActiveRecord
    module ConnectionAdapters
      module DatabaseLimitsExt
        def in_clause_length
          1000
        end
      end

      class AbstractAdapter
        prepend DatabaseLimitsExt
      end
    end
  end
end
```

This commit enables to configure `in_clause_length` value in the database.yml file.

* database.yml

```
production:
  adapter: mysql2
  in_clause_length: 1000
```

`in_clause_length` has been useful for Oracle enhanced adapter since Oracle database has a limit to allow the number of in list to 1000.
This parameter is not limited to MySQL. I have not found use cases for PostgreSQL and SQLite databases, though.

Skip this test if the `in_memory_db?` true since `run_without_connection` method executes `ActiveRecord::Base.remove_connection`.
SQLite 3 in memory database stores the tables inside the connection, once the conection is removed all of the tables under the connection will be lost.

yahonda commented:
Thanks for the review. I understand `in_clause_length` only supports preloading queries and does not support eager loading etc. Still, I think it is useful then let me leave this pull request open if other use cases are reported.

yahonda commented:
Closing this pull request. Will see if it will be useful with eager loading support.
